lib/web_ui/lib/src/engine/skwasm/skwasm_impl/memory.dart

```diff
@@ -0,0 +1,31 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'dart:ffi';
+import 'dart:js_interop';
+
+import 'package:ui/src/engine.dart';
+
+abstract class SkwasmObjectWrapper<T extends NativeType> {
+  Pointer<T> get handle;
+}
+
+typedef DisposeFunction<T extends NativeType> = void Function(Pointer<T>);
+
+class SkwasmFinalizationRegistry<T extends NativeType> {
+  SkwasmFinalizationRegistry(DisposeFunction<T> dispose)
+    : registry = DomFinalizationRegistry(((JSNumber address) =>
+      dispose(Pointer<T>.fromAddress(address.toDart.toInt()))
+    ).toJS);
+
+  final DomFinalizationRegistry registry;
+
+  void register(SkwasmObjectWrapper<T> wrapper) {
```

Comment:
One nice feature of `UniqueRef`/`CountedRef` is that they integrate with `Instrumentation` which helps you find leaks and double-frees. But maybe this can be added in a follow-up PR, if the design allows it.